packages/joint-core/src/dia/LinkView.mjs

```diff
@@ -1873,27 +1873,46 @@ export const LinkView = CellView.extend({
         var minDistance = Number.MAX_VALUE;
         var pointer = new Point(x, y);
 
-        viewsInArea.forEach(function(view) {
+        // filter out the current view
+        viewsInArea.filter((view) => view !== this).forEach(function(view) {
             const candidates = [];
+            const { model } = view;
             // skip connecting to the element in case '.': { magnet: false } attribute present
             if (view.el.getAttribute('magnet') !== 'false') {
-                candidates.push({
-                    bbox: view.model.getBBox(),
-                    magnet: view.el
-                });
+
+                if (model.isLink()) {
+                    const connection = view.getConnection();
+                    candidates.push({
+                        // find distance from the closest point of a link to pointer coordinates
+                        distance: connection.closestPoint(pointer).squaredDistance(pointer),
+                        magnet: view.el
+                    });
+                } else {
+                    candidates.push({
+                        // find distance from the center of the model to pointer coordinates
+                        distance: model.getBBox().center().squaredDistance(pointer),
+                        magnet: view.el
+                    });
+                }
             }
 
             view.$('[magnet]').toArray().forEach(magnet => {
+
+                const magnetBBox = view.getNodeBBox(magnet);
+                const rootCenter = model.isElement() ? model.getBBox().center() : view.getConnection().pointAt(0.5);
```

Comment:
This should be moved to `elementView` and `linkView`.  
For example, you could add a `getSnapPoint()` method to allow overriding the logic.